plugin/forward/forward_test.go

```diff
@@ -40,37 +38,148 @@ func TestList(t *testing.T) {
 }
 
 func TestSetTapPlugin(t *testing.T) {
-	input := `forward . 127.0.0.1
-	dnstap /tmp/dnstap.sock full
-	dnstap tcp://example.com:6000
-	`
-	stanzas := strings.Split(input, "\n")
-	c := caddy.NewTestController("dns", strings.Join(stanzas[1:], "\n"))
-	dnstapSetup, err := caddy.DirectiveAction("dns", "dnstap")
-	if err != nil {
-		t.Fatal(err)
-	}
-	if err = dnstapSetup(c); err != nil {
-		t.Fatal(err)
-	}
-	c.Dispenser = caddyfile.NewDispenser("", strings.NewReader(stanzas[0]))
-	if err = setup(c); err != nil {
-		t.Fatal(err)
-	}
-	dnsserver.NewServer("", []*dnsserver.Config{dnsserver.GetConfig(c)})
-	f, ok := dnsserver.GetConfig(c).Handler("forward").(*Forward)
-	if !ok {
-		t.Fatal("Expected a forward plugin")
-	}
-	tap, ok := dnsserver.GetConfig(c).Handler("dnstap").(*dnstap.Dnstap)
-	if !ok {
-		t.Fatal("Expected a dnstap plugin")
-	}
-	f.SetTapPlugin(tap)
-	if len(f.tapPlugins) != 2 {
-		t.Fatalf("Expected: 2 results, got: %v", len(f.tapPlugins))
-	}
-	if f.tapPlugins[0] != tap || tap.Next != f.tapPlugins[1] {
-		t.Error("Unexpected order of dnstap plugins")
+	for _, tt := range []struct {
+		name,
+		tapConfig string
+		assert func(t *testing.T, src *dnstap.Dnstap, actualTaps []*dnstap.Dnstap)
+	}{
+		{
+			name:      "one dnstap without message_types",
+			tapConfig: `dnstap tcp://example.com:6000`,
+			assert: func(t *testing.T, src *dnstap.Dnstap, actualTaps []*dnstap.Dnstap) {
+				if len(actualTaps) != 1 {
+					t.Fatalf("Expected: 1 results, got: %v", len(actualTaps))
```

Comment:
here and in other places, let's replace the `t.Fatalf()` with `t.Errorf(); return` to continue running other test cases